File: client-sdk/csharp/sdk/src/VertexCache/Sdk/Protocol/ResponseParserRegistry.cs
using System.Collections.Generic;
using VertexCache.Sdk.Protocol.Parsers;
using VertexCache.Sdk.Results;

namespace VertexCache.Sdk.Protocol
{
    public class ResponseParserRegistry
    {
        private readonly List<IResponseParser> _parsers = new()
        {
            new PongResponseParser(),
            new OkResponseParser(),
            new ErrorResponseParser(),
            new NilResponseParser(),
            new DeletedResponseParser(),
            new ValueResponseParser()
        };

        public VCacheResult Parse(string? response)
        {
            if (string.IsNullOrWhiteSpace(response))
                return VCacheResult.Failure(VCacheErrorCode.ProtocolError, "Empty response from server.");

            foreach (var parser in _parsers)
            {
                if (parser.CanParse(response))
                {
                    return parser.Parse(response);
                }
            }

            // Fallback: treat unknown + response as generic success
            if (response.StartsWith('+'))
                return VCacheResult.Success(response);

            return VCacheResult.Failure(VCacheErrorCode.Unknown, $"Unknown response: {response}");
        }
    }
}

File: client-sdk/go/sdk/protocol/response_parser_registry.go
package protocol

import (
	"strings"

	"vertexcache/sdk/protocol/parsers"
	"vertexcache/sdk/results"
)

type ResponseParser interface {
	CanParse(line string) bool
	Parse(line string) *results.VCacheResult
}

type ResponseParserRegistry struct {
	parsers []ResponseParser
}

func NewResponseParserRegistry() *ResponseParserRegistry {
	return &ResponseParserRegistry{
		parsers: []ResponseParser{
			&parsers.PongResponseParser{},
			&parsers.OkResponseParser{},
			&parsers.ErrorResponseParser{},
			&parsers.NilResponseParser{},
			&parsers.DeletedResponseParser{},
			&parsers.ValueResponseParser{},
		},
	}
}

func (r *ResponseParserRegistry) Parse(line string) *results.VCacheResult {
	for _, parser := range r.parsers {
		if parser.CanParse(line) {
			return parser.Parse(line)
		}
	}

	if strings.HasPrefix(line, "+") {
		return results.Success(line)
	}

	return results.Failure(results.Unknown, "Unknown response: "+line)
}
